Fix non thread safe state write in acmeda integration

atmurray opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The underlying module uses asyncio and the callback is run within the library's event loop.
As such, the following runtime error was being thrown when calling async_write_ha_state:

RuntimeError: Detected that integration 'acmeda' calls async_write_ha_state from a thread other than the event loop, which may cause Home Assistant to crash or data to corrupt. For more information, see https://developers.home-assistant.io/docs/asyncio_thread_safety/#async_write_ha_state at homeassistant/components/acmeda/base.py, line 65: self.async_write_ha_state(). 

To resolve this, the library is now initialised to use the existing event loop of home assistant. The callbacks are now async so that they are run within the home assistant event loop thereby allowing async_write_ha_state to be called.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #123509
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

atmurray committed:
Replaced callback to write_ha_state with schedule_update_ha_state
The underlying module uses asyncio and the callback is run within
the library's event loop.

atmurray commented:
This change was previously included in the pull request https://github.com/home-assistant/core/pull/124960 but has been split out into a dedicated pull request.

atmurray commented:
@bdraco would you mind checking this out.
The library already supports passing in an existing event loop so if I understand what you're saying the better fix would be to re-use the existing home assistant event loop in the library rather than creating a new one? If this is right, how do I get a reference to the home assistant event loop?

bdraco commented on homeassistant/components/acmeda/base.py:
```diff
@@ -62,7 +62,7 @@ async def async_will_remove_from_hass(self) -> None:
         """Write updated device state information."""
         LOGGER.debug("Device update notification received: %s", self.name)
-        self.async_write_ha_state()
+        self.schedule_update_ha_state()
```
```suggestion
        # The underlying library creates its own event loop that runs in a
        # separate thread. This is less than ideal, but is difficult to change
        # without a major refactor. Until that can be done, we must write
        # state with the thread-safe api.
        self.schedule_update_ha_state()
```

bdraco commented:
> @bdraco would you mind checking this out. The library already supports passing in an existing event loop so if I understand what you're saying the better fix would be to re-use the existing home assistant event loop in the library rather than creating a new one? If this is right, how do I get a reference to the home assistant event loop?

Yes that would be much better if possible.

`hass.loop` or `asyncio.get_running_loop()`

atmurray committed:
Change to use existing hass asyncio event loop in aiopulse module.
Change callback to be async so that it is added to
the event loop and not the executor.

atmurray commented:
> Yes that would be much better if possible.
> 
> `hass.loop` or `asyncio.get_running_loop()`

OK done, that was surprisingly easy. Only additional thing I had to do was to make the callback async so that it ran within the event loop.

joostlek commented on homeassistant/components/acmeda/base.py:
```diff
@@ -56,10 +56,10 @@ async def async_added_to_hass(self) -> None:
-        self.roller.callback_unsubscribe(self.notify_update)
+        self.roller.callback_unsubscribe(self.async_notify_update)
 
     @callback
```
If I remember correctly, callbacks can't be async

atmurray replied:
Within the library, the callback resolver looks at the callback function and if it is a coroutine (async) creates a task on the event loop, otherwise it runs the callback on the executor.

joostlek replied:
I'll wait for bdraco on this one, since the next patch will be out next friday, we don't have a rush to merge :)

bdraco replied:
This will work but its very inefficient since creating a task is an order of magnitude more latent than executing a callback.

In core we have a decorator, `@callback` to tell core how to run a callback.

If its decorated with `@callback` it runs in the loop, if not it runs in the executor

https://github.com/home-assistant/core/blob/54052792738e62366b2e07ad6dfe3b1d9049b0c1/homeassistant/core.py#L247
https://github.com/home-assistant/core/blob/54052792738e62366b2e07ad6dfe3b1d9049b0c1/homeassistant/core.py#L253
https://github.com/home-assistant/core/blob/54052792738e62366b2e07ad6dfe3b1d9049b0c1/homeassistant/core.py#L387

In the library you could do something similar, however if you are always running everything in the same loop there is likely no need to ever run anything in the executor and you could run the callbacks directly in the library and remove the executor code.

atmurray committed:
Merge branch 'dev' into acmeda-fixnotify

atmurray committed:
Merge branch 'dev' into acmeda-fixnotify

frenck commented:
This PR has a merge conflict, could you take a look @atmurray ?

../Frenck
